tests/test_FindTest.py

```diff
@@ -462,6 +462,9 @@ def test_find_citations(self):
             ('Id. foo', [id_citation('Id.,')]),
             # Reject citations that are part of larger words
             ('foo1 U.S. 1, 1. U.S. 1foo', [],),
+            # Long pin cite -- make sure no catastrophic backtracking in regex
+            ('1 U.S. 1, 2277, 2278, 2279, 2280, 2281, 2282, 2283, 2284, 2286, 2287, 2288, 2289, 2290, 2291',
+             [case_citation(metadata={'pin_cite': '2277, 2278, 2279, 2280, 2281, 2282, 2283, 2284, 2286, 2287, 2288, 2289, 2290, 2291'})]),
```

Comment:
Yeah, I dunno. If we accidentally reverted the regex to what it was, that line would take hours, not seconds, to complete. Don't really know how long, just that this will max out a CPU core seemingly indefinitely: `re.search(r'((, )?\d+)+ \(', '1 U.S. 1, 2277, 2278, 2279, 2280, 2281, 2282, 2283, 2284, 2286, 2287, 2288, 2289, 2290, 2291')`.

So we could throw in something to assert that say each call to `self.run_test_pairs` takes less than a second? But I'm less worried about reverting this exact thing and not noticing that the tests take forever, than I am about accidentally creating some other badly behaved regex later that we're not testing for. Don't have a great solution for that though.

In general for performance it would be cool if Github Actions had like a metrics server so tests could add a comment on how long they took to run relative to last time. I've thought about trying to jam that into a gist or something but I don't have a great solution for it.

Reply:
Sounds like we'd catch that eventually, at least for this one. There are a few things in the Github Marketplace related to benchmarks, but none that looked like a good fit. Closest thing was one that relied on pytest.